src/compressor/snappy/SnappyCompressor.h

```diff
@@ -61,12 +61,14 @@ class SnappyCompressor : public Compressor {
     dst.append(ptr, 0, sink.CurrentDestination()-ptr.c_str());
     return 0;
   }
-  virtual int decompress(bufferlist &src, bufferlist &dst) {
-    BufferlistSource source(src);
+  virtual int decompress(const bufferlist &src, bufferlist &dst) {
     size_t res_len = 0;
     // Trick, decompress only need first 32bits buffer
-    if (!snappy::GetUncompressedLength(src.get_contiguous(0, 8), 8, &res_len))
+    bufferlist tmp;
+    tmp.substr_of( src, 0, 4 );
```

Comment:
@yuyuyu101, no, it's not. The root cause is src.get_contiguous() call that may rebuild the src bufferlist and merge some initial segments into a single one. This invalidates an iterator stored in BufferlistSource before. To fix that temporary bufferlist named tmp has been introduced and non-destructive substr_of used instead of get_contiguous().

The rationale to replace 8 with 4 ones is simply to match the comment in the code and actual GetUncompressedLength needs - it looks like it needs 32 bits only.

Besides I believe that compressor shouldn't modify input buffer thus I made some refactoring to prohibit that by making this parameter as const.